crates/katana/rpc/rpc/src/torii.rs

```diff
@@ -0,0 +1,204 @@
+
+                    let block_transactions = block_transactions
+                        .into_iter()
+                        .map(|tx| {
+                            let receipt = ReceiptBuilder::new(tx.hash, provider)
+                                .build()
+                                .expect("Receipt should exist for tx")
+                                .expect("Receipt should exist for tx");
+                            (tx, MaybePendingTxReceipt::Receipt(receipt))
+                        })
+                        .collect::<Vec<_>>();
+
+                    // Add transactions to the total and break if MAX_PAGE_SIZE is reached
+                    for transaction in block_transactions {
+                        transactions.push(transaction);
+                        if transactions.len() >= MAX_PAGE_SIZE {
+                            next_cursor.block_number = block_number;
+                            next_cursor.transaction_index = MAX_PAGE_SIZE as u64;
+                            return Ok(TransactionsPage { transactions, cursor: next_cursor });
+                        }
+                    }
+                }
+            }
+
+            if let Some(pending_state) = this.sequencer.pending_state() {
+                let remaining = MAX_PAGE_SIZE - transactions.len();
+
+                // If cursor is in the pending block
+                if cursor.block_number == latest_block_number + 1 {
+                    let mut pending_transactions = pending_state
+                        .executed_txs
+                        .read()
+                        .iter()
+                        .skip(cursor.transaction_index as usize)
+                        .take(remaining)
+                        .map(|(tx, info)| {
+                            (
+                                tx.clone(),
+                                MaybePendingTxReceipt::Pending(PendingTxReceipt::new(
+                                    tx.hash,
+                                    info.receipt.clone(),
+                                )),
+                            )
+                        })
+                        .collect::<Vec<_>>();
+
+                    // If there are no transactions after the index in the pending block
+                    if pending_transactions.is_empty() {
+                        // Wait for a new transaction to be executed
+                        let maybe_listener = {
+                            let inner = this.sequencer.block_producer().inner.read();
+                            if let BlockProducerMode::Interval(block_producer) = &*inner {
+                                Some(block_producer.add_listener())
+                            } else {
+                                None
+                            }
+                        };
+
+                        if let Some(mut rx) = maybe_listener {
+                            pending_transactions = futures::executor::block_on(rx.next())
```

Comment:
Forgot to mention, but this would block the thread when waiting for txs to be available in the channel. We should change this to using async channel, and awaiting on it outside of the current blocking context.

This is what caused not being able to shutdown while a long polling is happening.